atc/worker/volume.go

```diff
@@ -104,7 +108,17 @@ func (v *volume) Destroy() error {
 	return v.bcVolume.Destroy()
 }
 
-func (v *volume) COWStrategy() baggageclaim.COWStrategy {
+// NOTE (runtime/#4964): in Linux kernel 5.x and greater,
+// We lose the ability the make a COW of a COW. Overlay will return an error.
```

Comment:
Hey,

Thanks for the PR!

> in Linux kernel 5.x and greater, we lose the ability the make a COW of a COW.
> Overlay will return an error.

reading through the details of how `overlayfs` works (see [overlayfs entry]), it
turns out that this is not really true (I originally thought it was, but there's
a catch), what can't be done is to have a `cow` of a `cow` where the `lowerdir`
of the "cow-cow" references an `upperdir` used somewhere else.

[overlayfs entry]: https://www.kernel.org/doc/Documentation/filesystems/overlayfs.txt

	FAIL

		mount \
			-t overlay \
			cow1 \
			-o lowerdir=./cow1-lower,upperdir=./cow1-upper,workdir=./cow1-work \
			./cow1-merged

		mount \
			-t overlay \
			cow1-cow \
			-o lowerdir=./cow1-upper,upperdir=./cow1-cow-upper,workdir=./cow1-cow-work \
			./cow1-cow-merged   |
					    |
					    '--> using an upper: PROBLEM!


	SUCCESS

		mount \
			-t overlay \
			cow1 \
			-o lowerdir=./cow1-lower,upperdir=./cow1-upper,workdir=./cow1-work \
			./cow1-merged

		mount \
			-t overlay \
			cow1-cow \
			-o lowerdir=./cow1-merged,upperdir=./cow1-cow-upper,workdir=./cow1-cow-work \
			./cow1-cow-merged   |
					    | 
					    '--> not an upper: just fine :)


I think we might be able to fix the problem that we currently have by changing
the way that we deal with ancestors :thinking:

I'll first give a try at changing the code in baggageclaim to see if I can get
anywhere, then come back to this one :+1: I'm 99% sure we can get this fixed
over there with very little code changes.

Thanks!

Reply:
created https://github.com/concourse/baggageclaim/pull/35

Reply:
SMART

Reply:
should I close this pr?

Reply:
yeaah 😁  thanks for the PR, though!